File: threedollar-push/src/main/java/com/depromeet/threedollar/push/listener/SlackNotificationEventListener.java
package com.depromeet.threedollar.push.listener;

import com.depromeet.threedollar.common.model.event.ApplicationStateChangedEvent;
import com.depromeet.threedollar.common.model.event.ServerExceptionOccurredEvent;
import com.depromeet.threedollar.infrastructure.external.client.slack.SlackWebhookApiClient;
import com.depromeet.threedollar.infrastructure.external.client.slack.dto.request.PostSlackMessageRequest;
import lombok.RequiredArgsConstructor;
import org.springframework.context.event.EventListener;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Component;

import static com.depromeet.threedollar.common.type.template.SlackMessageTemplateType.ERROR_MESSAGE;
import static com.depromeet.threedollar.common.type.template.SlackMessageTemplateType.INFO_MESSAGE;

@RequiredArgsConstructor
@Component
public class SlackNotificationEventListener {

    private final SlackWebhookApiClient slackNotificationApiClient;

    @Async
    @EventListener
    public void sendErrorNotification(ServerExceptionOccurredEvent event) {
        slackNotificationApiClient.postMonitoringMessage(PostSlackMessageRequest.of(ERROR_MESSAGE.generateMessage(
            event.getApplicationType().getDescription(),
            event.getErrorCode().getCode(),
            event.getRequestUri(),
            event.getException(),
            event.getTimeStamp(),
            event.getErrorCode().getMessage(),
            event.getUserMetaValue()
        )));
    }

    @Async
    @EventListener
    public void sendInfoNotification(ApplicationStateChangedEvent event) {
        slackNotificationApiClient.postMonitoringMessage(PostSlackMessageRequest.of(INFO_MESSAGE.generateMessage(
            event.getApplicationName(),
            event.getApplicationUid(),
            event.getMessage(),
            event.getTimeStamp()
        )));
    }

}

File: threedollar-api-boss/src/main/kotlin/com/depromeet/threedollar/api/boss/listener/slack/SlackNotificationEventListener.kt
package com.depromeet.threedollar.api.boss.listener.slack

import java.util.stream.Collectors
import org.springframework.context.event.EventListener
import org.springframework.scheduling.annotation.Async
import org.springframework.stereotype.Component
import com.depromeet.threedollar.common.model.event.ApplicationStateChangedEvent
import com.depromeet.threedollar.common.model.event.ServerExceptionOccurredEvent
import com.depromeet.threedollar.domain.mongo.boss.domain.category.BossStoreCategoryRepository
import com.depromeet.threedollar.domain.mongo.boss.event.registration.NewBossAppliedRegistrationEvent
import com.depromeet.threedollar.external.client.slack.SlackWebhookApiClient
import com.depromeet.threedollar.external.client.slack.dto.request.PostSlackMessageRequest
import com.depromeet.threedollar.external.client.slack.type.SlackNotificationMessageType
import com.depromeet.threedollar.external.client.slack.type.SlackNotificationMessageType.NEW_BOSS_REGISTRATION_MESSAGE

@Component
class SlackNotificationEventListener(
    private val slackNotificationApiClient: SlackWebhookApiClient,
    private val bossStoreCategoryRepository: BossStoreCategoryRepository
) {

    @Async
    @EventListener
    fun sendRegistrationNotification(event: NewBossAppliedRegistrationEvent) {
        slackNotificationApiClient.postBossManagerMessage(
            PostSlackMessageRequest.of(
                NEW_BOSS_REGISTRATION_MESSAGE.generateMessage(
                    event.registration.id,
                    event.registration.boss.name,
                    event.registration.boss.socialInfo.socialType,
                    event.registration.boss.businessNumber.getNumberWithSeparator(),
                    event.registration.store.name,
                    bossStoreCategoryRepository.findAllCategoriesByIds(event.registration.store.categoriesIds).stream().map { it.name }.collect(Collectors.joining(", ")),
                    event.registration.store.contactsNumber.getNumberWithSeparator(),
                    event.registration.store.certificationPhotoUrl
                )
            )
        )
    }

    @Async
    @EventListener
    fun sendErrorNotification(event: ServerExceptionOccurredEvent) {
        slackNotificationApiClient.postMonitoringMessage(
            PostSlackMessageRequest.of(
                SlackNotificationMessageType.ERROR_MESSAGE.generateMessage(
                    event.applicationType.description,
                    event.errorCode.code,
                    event.requestUri,
                    event.exception,
                    event.timeStamp,
                    event.errorCode.message,
                    event.userMetaValue,
                )
            )
        )
    }

    @Async
    @EventListener
    fun sendInfoNotification(event: ApplicationStateChangedEvent) {
        slackNotificationApiClient.postMonitoringMessage(
            PostSlackMessageRequest.of(
                SlackNotificationMessageType.INFO_MESSAGE.generateMessage(
                    event.applicationType.description,
                    event.applicationUid,
                    event.message,
                    event.timeStamp
                )))
    }

}
